crates/install-wheel-rs/src/linker.rs

```diff
@@ -280,53 +280,93 @@ fn clone_wheel_files(
-            for entry in fs::read_dir(from)? {
-                clone_recursive(site_packages, wheel, &entry?)?;
+    match attempt {
+        CloneAttempt::First => {
+            *attempt = CloneAttempt::Subsequent;
+            if let Err(err) = reflink_copy::reflink(&from, &to) {
+                if err.kind() == std::io::ErrorKind::AlreadyExists {
+                    // If copying fails and the directory exists already, it must be merged recursively.
+                    if entry.file_type()?.is_dir() {
+                        for entry in fs::read_dir(from)? {
+                            clone_recursive(site_packages, wheel, &entry?, attempt)?;
+                        }
+                    } else {
+                        // If file already exists, overwrite it.
+                        let tempdir = tempdir_in(site_packages)?;
+                        let tempfile = tempdir.path().join(from.file_name().unwrap());
+                        reflink_copy::reflink(from, &tempfile)?;
+                        fs::rename(&tempfile, to)?;
+                    }
+                } else {
+                    // switch to copy fallback
+                    *attempt = CloneAttempt::UseCopyFallback;
+                    return clone_recursive(site_packages, wheel, entry, attempt);
+                }
             }
-        } else {
-            // If file already exists, overwrite it.
-            let tempdir = tempdir_in(site_packages)?;
-            let tempfile = tempdir.path().join(from.file_name().unwrap());
-            reflink_copy::reflink(from, &tempfile)?;
-            fs::rename(&tempfile, to)?;
         }
-    } else {
-        // Other errors should be tracked
-        reflink.map_err(|err| Error::Reflink {
-            from: from.clone(),
-            to: to.clone(),
-            err,
-        })?;
+        CloneAttempt::Subsequent => {
+            if let Err(err) = reflink_copy::reflink(&from, &to) {
+                if err.kind() == std::io::ErrorKind::AlreadyExists {
+                    // If copying fails and the directory exists already, it must be merged recursively.
+                    if entry.file_type()?.is_dir() {
+                        for entry in fs::read_dir(from)? {
+                            clone_recursive(site_packages, wheel, &entry?, attempt)?;
+                        }
+                    } else {
+                        // If file already exists, overwrite it.
+                        let tempdir = tempdir_in(site_packages)?;
+                        let tempfile = tempdir.path().join(from.file_name().unwrap());
+                        reflink_copy::reflink(from, &tempfile)?;
+                        fs::rename(&tempfile, to)?;
+                    }
+                } else {
+                    return Err(Error::Reflink { to, from, err });
+                }
+            }
+        }
+        CloneAttempt::UseCopyFallback => {
+            fs::copy(&from, &to)?;
```

Comment:
Does this need to recursively copy? Does it work if `from` is a directory? (I notice that in `copy_wheel_files`, we use `walkdir`.)

Reply:
This does need to recursively copy, my mistake.